monitoring: Metrics Monitoring Feature

R2ZER0 opened the pull request:
This PR creates a new 'monitoring' module within the django app, to
add the metric monitoring feature.

This feature:
* Allows storing a history of snapshots of metrics (calculated
values) for each Publisher, Funder and SourceFile.
* Provides a management command to calculate and store a snapshot.
* Provides an API to expose these snapshots publicly.

R2ZER0 committed:
monitoring: Metrics Monitoring Feature

This commit creates a new 'monitoring' module within the django app, to
add the metric monitoring feature.

This feature:
* Allows storing a history of snapshots of metrics (calculated
values) for each Publisher, Funder and SourceFile.
* Provides a management command to calculate and store a snapshot.
* Provides an API to expose these snapshots publicly.

R2ZER0 committed:
data_quality: Add --threads opt to rewrite_quality_data

Allow changing the number of threads when generating quality data (my
laptop can't handle the default 8).

R2ZER0 commented:
Reminder: We'll also need to add the gather metrics management command to the daily `data_run.sh`

michaelwood commented on datastore/monitoring/api.py:
```diff
@@ -0,0 +1,89 @@
+)
+
+
+class SnapshotAPIView(ListAPIView):
```
The APIViews "belongs" in `/datastore/api/<what they export/logically do>/api.py` . This collates all the code that runs those "exporting" data out of the datastore into one place. This is an arbitrary decision but is important to stay consistent with that to avoid design fragmentation.

e.g.
```
$ git grep APIView
datastore/api/dashboard/api.py:33:class Publishers(generics.ListAPIView):
datastore/api/dashboard/api.py:53:class Publisher(generics.RetrieveAPIView):
datastore/api/dashboard/api.py:62:class Sources(generics.ListAPIView):
datastore/api/experimental/api.py:13:class CurrentLatestGrants(generics.ListAPIView):
datastore/api/org/api.py:17:class OrganisationListView(generics.ListAPIView):
datastore/api/org/api.py:35:class OrganisationDetailView(generics.RetrieveAPIView):
datastore/api/org/api.py:58:        Based on upstream DRF GenericAPIView.get_object().
datastore/api/org/api.py:103:class OrganisationGrantsMadeView(generics.ListAPIView):
datastore/api/org/api.py:134:class OrganisationGrantsReceivedView(generics.ListAPIView):
datastore/api/org/api_schema.py:1:from drf_spectacular.views import SpectacularAPIView
datastore/api/org/api_schema.py:35:class APISchemaView(SpectacularAPIView):
```

R2ZER0 replied:
Moved it into `api/monitoring/api.py`

michaelwood commented on datastore/monitoring/api.py:
```diff
@@ -0,0 +1,89 @@
+
+    def get_queryset(self):
+
+        # If no date is specified, use today's date
```
minor - comment might be redundant as you've actually implemented that below?

michaelwood commented on datastore/monitoring/api.py:
```diff
@@ -0,0 +1,89 @@
+
+        ident_filter = {idf: OuterRef(idf) for idf in self.identifier_fields}
+
+        # This is not the most efficient query, but it's only meant to be accessed occasionally, and still renders in ms
```
minor - A description of what it does in a comment might be good instead here!

michaelwood commented on datastore/monitoring/models.py:
```diff
@@ -0,0 +1,72 @@
+
+
+class PublisherMetricsRecord(models.Model):
+    timestamp = models.DateTimeField()
```
I'm just wondering if there is a need to be able to group datasets/snapshots in a way that doesn't rely on datetime, how difficult would it be to remove an erroneous one later on if two were generated at the same time?

R2ZER0 replied:
That is a good point, especially since snapshots timestamps are the `.datetime` of the last GetterRun, so if we created 2+ snapshots of the same GetterRun, they'd all have the same timestamp.

I guess there's the implicit Django `.id` autoincrement, do we want to make it explicit like GetterRuns?

michaelwood replied:
(in a meeting we discussed an additional model that groups snapshots together for convenience)

michaelwood commented on datastore/data_quality/management/commands/rewrite_quality_data.py:
```diff
@@ -82,7 +88,7 @@ def handle(self, *args, **options):
                 )
 
-            with Pool(8) as process_pool:
+            with Pool(options.get("threads") or 8) as process_pool:
```
minor - can drop the `or 8` as you've set a default.

michaelwood reviewed (changes requested):
This all looks good, only one thing I think I'd like to see change before merging which is the placement of the APIViews not being in the same module/app as their corresponding urls.

michaelwood commented:
Great to see all the hard work on the testing. Looks like coverage indicator is looking good. https://coveralls.io/github/ThreeSixtyGiving/datastore?branch=rg/publisher-etc-monitoring

michaelwood commented:
One other little request is to add something to the README in the key modules section. Thanks

R2ZER0 committed:
readme: Add info about monitoring module

R2ZER0 committed:
monitoring: Test Monitoring Metrics & Queries

This commit adds further tests for the monitoring module:
* Test multiple GetterRuns in one day
* Test removing a Funder
* Test adding a Publisher
* Test SourceFile downtime
* fake_testdata module to generate on-the-fly fake GetterRuns etc

As well as fixing bugs found by these tests.

R2ZER0 committed:
monitoring: Add helper management commands

and rename the create_monitoring_snapshot command for consistency.

R2ZER0 committed:
monitoring: Expand Dataset Metrics

to include headline figures from the data quality dashboard.

R2ZER0 committed:
monitoring: Minor fixes

Fix type annotations

michaelwood commented on requirements_dev.txt:
```diff
@@ -5,9 +5,9 @@
 #
 -e git+https://github.com/ThreeSixtyGiving/datagetter.git@b907c5c0463c853b56abe4ca52df88d285cc7f67#egg=datagetter
-    # via -r requirements.in
+    # via -r /home/rikki/Projects/datastore/requirements.in
```
pip-compile strikes again, can you remove the full path.

michaelwood commented on datastore/data_quality/management/commands/rewrite_quality_data.py:
```diff
@@ -82,7 +88,7 @@ def handle(self, *args, **options):
                 )
 
-            with Pool(8) as process_pool:
+            with Pool(options.get("threads")) as process_pool:
```
I think `options.threads` here? Pool can't take None and it's been initialised through through the default

michaelwood commented on datastore/monitoring/models.py:
```diff
@@ -0,0 +1,169 @@
+                )
+            )
+            .values_list("latest_record", flat=True)
+        )
```
No change needed, just a thought  - Appreciate the comments here for documentation, though one thing that I've found helpful with more complicated queries is just an example of what   `latest_record_pks`  could look like (unless it's too huge to describe) so for other things I've done something like: 
```
# Generates latest_records_pk
[ 
  { "blah": "cat", "moo" : [ "Labrador" ,  "Alsatian" ] } 
  ...
]
```

I think if anyone was working on this the first thing they'd do was `print latest_record_pks`, examples I think can be worth a 1000 words!

R2ZER0 replied:
That's a point, `latest_record_pks` is literally just a list of integer primary keys `[123, 456, 73737, ...]` of the *MetricsRecords to display, but i should make that clear. It makes more sense if you see the SQL that's generated, i can include that too

michaelwood reviewed (approved):
Great - Approved, a couple of possible things to follow up with.
